WIP: better utilize the MillController

Leptopoda opened the pull request:
It makes the controller a singleton and moves the recorder and tip into it.

It also splits the game page into more widgets making it more performant and less error-prone. The tip is now handled in its separate value listenable.

@calcitem this commit isn't the entire migration, but it will eventually fix several (if not all) issues introduced in the last PR. This significant restructure was needed to achieve this.

To continue, I need some input from you:
1. can you please explain what `engineToGo` exactly does? This is the function that blocks fixing the above-mentioned errors.
2. it seems like you pass the engine (cpp) the fen representation of the board on each move. But you only pass it `recorder.lastPositionWithRemove`. This is not always the same as the fen representation of the current board. The most recent one could be obtained by `position._fen`. I've played a game and logged both versions:

```
I/flutter ( 1413): debug old: ********/********/******** w p p 0 9 0 9 0 0 1
I/flutter ( 1413): debug fen: ********/********/***O**** b p p 1 8 0 9 0 1 1
I/flutter ( 1413): debug old: ********/********/******** w p p 0 9 0 9 0 0 1
I/flutter ( 1413): debug fen: ********/@*******/**OO**** b p p 2 7 1 8 0 1 2
I/flutter ( 1413): debug old: ********/******@*/*OOO**** b p p 3 6 1 7 0 1 3
I/flutter ( 1413): debug fen: ********/******@*/*OOO**** b p p 3 6 1 7 0 1 3
I/flutter ( 1413): debug old: ********/******@*/*OOO**** b p p 3 6 1 7 0 1 3
I/flutter ( 1413): debug fen: ********/****@*@*/OOOO**** b p p 4 5 2 6 0 1 4
I/flutter ( 1413): debug old: ********/****@*@*/OOOO***O b p p 5 4 2 5 0 1 6
I/flutter ( 1413): debug fen: ********/****@*@*/OOOO***O b p p 5 4 2 5 0 1 6
I/flutter ( 1413): debug old: ********/****@*@*/OOOO***O b p p 5 4 2 5 0 1 6
I/flutter ( 1413): debug fen: ********/**@*@*@*/OOOO**OO b p p 6 3 3 4 0 1 7
I/flutter ( 1413): debug old: ********/@*@*@***/OOOO*OOO b p p 7 2 3 3 0 1 8
I/flutter ( 1413): debug fen: ********/@*@*@***/OOOO*OOO b p p 7 2 3 3 0 1 8
I/flutter ( 1413): debug old: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug fen: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug old: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug fen: *****O**/@*@*@*@*/OOOOOOOO b p p 9 0 4 1 0 1 11
I/flutter ( 1413): debug old: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug fen: ****O*@*/@*@*@*@*/OOOOOOOO b m s 9 0 5 0 0 3 12
I/flutter ( 1413): debug old: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug fen: ***O**@*/@*@*@**@/OOOOOOOO b m s 9 0 5 0 0 8 14
I/flutter ( 1413): debug old: ********/@*@***@*/OOOOOOOO b p p 8 1 3 2 0 1 10
I/flutter ( 1413): debug fen: ***O**@*/@@**@**@/OOOOOOOO b m r 9 0 5 0 1 10 14
I/flutter ( 1413): debug old: ******@*/@@**@**@/OOOOOOOO w m s 8 0 5 0 0 0 15
I/flutter ( 1413): debug fen: ******@*/@@O*@**@/OO*OOOOO b m s 8 0 5 0 0 3 16
I/flutter ( 1413): debug old: ******@*/*@**@*@*/OOOOOOOO b m s 8 0 4 0 0 1 18
I/flutter ( 1413): debug fen: ******@*/*@**@*@*/OOOOOOOO b m s 8 0 4 0 0 1 18
I/flutter ( 1413): debug old: ******@*/*@**@*@*/OOOOOOOO b m s 8 0 4 0 0 1 18
I/flutter ( 1413): debug fen: ******@*/O*@*@*@*/*OOOOOOO b m s 8 0 4 0 0 6 20
I/flutter ( 1413): debug old: ******@*/****@*@*/OOOOOOOO b m s 8 0 3 0 0 1 22
I/flutter ( 1413): debug fen: ******@*/****@*@*/OOOOOOOO b m s 8 0 3 0 0 1 22
I/flutter ( 1413): debug old: ******@*/****@*@*/OOOOOOOO b m s 8 0 3 0 0 1 22
I/flutter ( 1413): debug fen: @*****@*/O*****@*/*OOOOOOO b m s 8 0 3 0 0 6 23
I/flutter ( 1413): debug old: ******@*/****@*@*/OOOOOOOO b m s 8 0 3 0 0 1 22
I/flutter ( 1413): debug fen: @*****@*/O*O*****/@O*OOOOO b m s 8 0 3 0 0 11 25
I/flutter ( 1413): debug old: ******@*/****@*@*/OOOOOOOO b m s 8 0 3 0 0 1 22
I/flutter ( 1413): debug fen: @*******/O*O*O***/@O@O*OOO b m s 8 0 3 0 0 16 26
```

In this log, the "old" one is the fen stored in the recorder, and the "fen" is the one stored in position. You can clearly see that the old one lacks behind quite a bit (it only gets updated when a piece is being removed). IMO the engine should always have the most recent one. Is this right?

Leptopoda commented:
ok seems like we are adding them back eventually. I'll have a look again.

``` dart
    final moves = position._movesSinceLastRemove;

    [...]

    if (moves != null) {
      posFenStr.write(" moves $moves");
    }
```

calcitem commented:
The purpose of `engineToGo` is to send FEN (got by `_getPositionFen`) to the engine and then send the `go` command to the engine to search for the best move.

The sent FEN is `$startPosition moves $_movesSinceLastRemove` because the engine will consider that the same position will lead to a draw when searching, thus evaluating the position's value more accurately. So simply sending `startPosition` (`position._fen`) is not sufficient.

When `_MoveType` is `_MoveType.move`, `lastPositionWithRemove = position._fen; `

The reference source is https://github.com/hezhaoyun/chessroad/blob/master/lib/routes/battle-page.dart#L96

ChessRoad can be downloaded from https://play.google.com/store/apps/details?id=cn.apppk.apps.chessroad&hl=de&gl=DE

Leptopoda commented:
Hi @calcitem 
I just fixed a few bugs.
Can you please test a bit and confirm they've been closed?

Also a review of the current status would be great :upside_down_face:

calcitem commented:
Thank you for your hard work at Christmas!

I will test it in six hours.

Leptopoda commented:
Ok currently the gesturedetector is not 100% responsive...

I'll fix that later.

calcitem commented:
Sorry, I just installed the new apk today and haven't had time to verify it in detail. 

I had a brief look at the GUI. There are some changes. I think we can clarify the direction of subsequent GUI improvements at https://github.com/calcitem/Sanmill/issues/296. For example, my idea is that we need to remove a horizontal line in the Header  (this horizontal line is directly following the ChessRoad app). In this case, when the board moves, it is okay that the top Header does not move downward along with it. But if the horizontal line is not deleted, I think it is not beautiful enough.

calcitem commented:
Issues:

- [ ] When adjust the width of piece, error: 'package:flutter/src/material/slider.dart': Failed assertion: line 147 pos 15: 'value >= min && value <= max': is not true.

Editing...

calcitem commented:
Part of the test results was temporarily replied. There are some problems I don't remember if and if I have created a new issue.

Leptopoda commented:
> It is recommended to rename MillController to GameController. It should not be coupled with Mill.

I don't think this is needed. The controller is deeply integrated with Mill in general and as there can't be any further abstraction (for now) the name is pretty accurate.

> At the end of the refactoring, it is best that there are only logic implementations related to mill in the mill folder, and files similar to tap_handler.dart are recommended to be placed elsewhere.

There already is a todo ;) 

> After the Catcher page appears once, a red page will appear again, which has not been encountered before. Not yet reproduced.

Probably an issue with catcher. When did this happen?

> The font size setting is currently not working.

I've already said a few times: it is working but not reactive. You'll need to restart the app for that. There already is an issue for that.

> The edge of the piece is now fixedly colored black. The edge color was the median value of the two sides' colors in the past, so the difference would not look too big. This problem can be solved later, first record it.

What do you mean? I think the shadow color should be the same as in reality the shadow is always black no matter what color the piece is (as long as the surface isn't shiny but let's not get into physics 🤪). So what do you want? We can also implement custom colors for each piece.

> Writes the given number to the buffer. it will add an extra space in front of single-digit numbers but won't fix three-digit cases. It doesn't seem to work. Numbers <10 and >10 are aligned in the center in the move list.

Fix is coming👍

> During the startup of the App, a rotating progress bar will appear in the middle of the screen, and there will be a noticeable delay.

Yes. There isn't anything we can do about this. We need to await a few things mainly to startup the native engine (which is asynchronous) and the audio service (also async). We can't start the game without those being initialized. We might wanna change the ui or do this directly during startup so it wont appear to the user that much. 
I'm no UX guy so I'll leave it up to you!

> It is best to draw notation after the board line is drawn, not to overwrite it after adjusting it to a significant value.

What do you mean? I don't understand.

> In the initial state, the size of the pieces is small. When adjust the width of piece, error: 'package:flutter/src/material/slider.dart': Failed assertion: line 147 pos 15: 'value >= min && value <= max': is not true.

> When taken back, a red circle will be left on the board. And take back is currently not working correctly. It may be because there is something wrong with the move list.

Boht related and a todo is in the code. Should be a fairly easy fix.

> The spacing between navigation toolbar and board changes from 2px to 1px.

Didn't change anything. I'll have a look.

> Change the value of Animation duration to a significant value. Then when the drawing is not completed, the left menu is opened, and an error occurs: _BoardState#e4119(ticker active) was disposed with an active Ticker. _BoardState created a Ticker via its SingleTickerProviderSt.

That's going to be tricky. I'll see what I can do. Probably I need to better dispose the animation controller.

> Tip no longer changes, always showing Welcome.

That's a bummer. Fix will come.

> The background color of the navigation menu icon in the upper left corner changes.

What do you mean?

> Notations 1 and g are not be displayed on the board. Only 234567 and abcdef are displayed.

Lul. Fix will come.

> Move list has not been clear when starting a new game. So cannot test importing move list twice.

There are quite a few issues with starting and ending the game... Probably the last to be fixed xD



I need input on quite a few of them. Please open a new issue for them. Everything I didn't ask for more input does not need a new issue. Thanks for the debug work.

calcitem commented:
Partly replied.

Leptopoda commented:
Hi @calcitem,
i've rewritten quite a big chunk of code.

Every major bug should be fixed now. Wherever the `engineToGo` is used still needs some work (this means where the ai needs to moves first).

You might also have noticed that I started writing tests for the bugs we closed so they wont reappear.

I'm also going to rewrite the import export service as there are quite a few duplications. To actually be sure I don't' break anything I'd like to cover it with tests before I start so I can verify the result in the end. Because I don't have all variations of the supported notation styles I rely on you sending me a few examples.

calcitem commented:
Thank you very much! The infrastructure becomes more elegant!

I will try to run the test, if I can, I can add several test cases directly.

Leptopoda commented:
Just fyi there are a few bugs related to the new game recorder (mainly the export and the moves since last remove are not functioning propperly) I already fixed them and added tests and will push them when I'm home again.

So no need to report issues with that ;)
Also as a result of that the game history navigation might be broken as well.

calcitem commented:
Understood.  I will add test cases first, so I don't have to worry about the test results.  Ha ha.

calcitem commented on src/ui/flutter_app/lib/services/mill/src/engine.dart:
```diff
@@ -220,3 +214,97 @@ class NativeEngine extends Engine {
+  aiVsAi,
+
+  /// Not Implemented
+  humanVsCloud,
```
I once considered interacting with http://ninemensmorris.ist.tugraz.at:8080/ through HTTP requests, but I never got a response after sending an email to this university. I'm not sure if it can be called. If it is not suitable, the remaining code can be deleted later.

calcitem commented on src/ui/flutter_app/lib/services/mill/src/history_navigation.dart:
```diff
@@ -156,3 +156,76 @@ class HistoryNavigator {
 }
+
+enum HistoryMove { forwardAll, backAll, forward, backN, backOne }
+
```
How about renaming `forward` to `forwardOne`?

Leptopoda commented on src/ui/flutter_app/lib/services/mill/src/recorder.dart:
```diff
@@ -21,7 +21,18 @@ part of '../mill.dart';
-  // TODO: [Leptopoda] as this is not related to the recorder in general we should probably move this to the import export service.
+  @override
+  String toString() {
+    final buffer = StringBuffer("[ ");
```
The `toString` method is nothing we'll use for the game. The toString of a list is commonly represented with square brackets and as this is a list (PointedList) we will print it as such. 

The method is only used when you want to print out the contents of the list and we either use this in logs (not now) but I use it to verify the test result of the import as we otherwise have no way to verify that the list has been popilated.

Leptopoda commented on src/ui/flutter_app/test/helpers/test_mills.dart:
```diff
@@ -1,25 +1,98 @@
+  String get exportNotStandard;
+}
+
+class WinnGame extends TestGame {
```
My thinking was that we'll have a few scenarios to test. For example the export you see in there was where the player one has won the game because the other player had less than three pieces.

Therefore we can't test the `movesSinceLastRemove` (as the last remove was the last move) nor can we test other GameOverReasons. The idea is that we have a few classes with different games and iterate over them to verify multiple cases.

I named it WinnGame as it has a recorded game (the export) and a few other metadata about it to verify our code.

How would you name it? Maybe something like WinnLessThanThreeGame ???

calcitem replied:
Maybe it is OK. Win, not Winn.

Leptopoda commented on src/ui/flutter_app/test/helpers/test_mills.dart:
```diff
@@ -1,25 +1,98 @@
+      '[ (3,4), (2,7), (2,4), (2,3), (3,5), (2,1), (3,6), -(2,3), (2,3), (3,3), (2,5), (3,2), -(2,3), (2,3), (1,6), (1,8), (2,8), (3,8), (1,5), (1,4), (1,6)->(1,7), (2,5)->(2,6), (3,5)->(2,5), (1,4)->(1,3), (2,5)->(3,5), -(2,6), (1,8)->(1,1), (2,4)->(2,5), -(2,1), (2,7)->(3,7), (1,7)->(2,7), (2,3)->(2,4), (3,2)->(3,1), (1,3)->(2,3), (1,5)->(1,4), (1,1)->(2,1), (1,4)->(1,3), (2,1)->(2,2), -(2,7),  ...
+
+  @override
+  String get exportNotStandard => ' 1. (3,4)  2. (2,7)\n'
```
So we could remove the ninechess export?

So we only keep it for internal use?

calcitem replied:
I will finally want to remove it, but because the move list exported to the PlayOK format has not yet been implemented, if it is removed directly, the option to export to the standard format notation is meaningless.  

We can consider not testing, but the function will be retained for now. What do you think?

Leptopoda replied:
> the option to export to the standard format notation is meaningless.

then we should just hide it until we have the other thing implemented :woman_shrugging:

calcitem replied:
OK!

Leptopoda commented on src/ui/flutter_app/lib/services/mill/src/tap_handler.dart:
```diff
@@ -84,53 +81,41 @@ class TapHandler {
-                    controller.position.pieceOnBoardCount[us] == 3)) {
+                (pieceOnBoardCount == DB().rules.flyPieceCount ||
+                    pieceOnBoardCount == 3)) {
+              // TODO: [Calcitem, Leptopoda] why is the [DB().rules.flyPieceCount] not respected?
```
In this case a less than operation might make the code more readable.

calcitem replied:
Yes!

calcitem replied:
I remembered one thing. If it is modified, it needs to be verified whether it can be handled correctly when the number of stones is less than 3. It is expected that the game is over and not here.  

By the way, much of the logic of the position module still uses the basic logic of NineChess. There are many if and else branches. I don't know if there is a better way to re-implement it, perhaps using a state machine?  

However, the reconstruction of these functions does require careful testing with test cases, and the workload is relatively large.

Leptopoda commented:
@calcitem 
Can you please give me a list of stuff that still needs to be fixed (except the AiVsAi)?

calcitem commented:
Okay, I will return to test it, and I expect to provide the results in a day.

Does the throw want to achieve an effect similar to Qt's emit?

calcitem commented:
How about https://pub.dev/packages/bloc ?

Is it useful for us?

Leptopoda commented:
> Okay, I will return to test it, and I expect to provide the results in a day.
> 
> Does the throw want to achieve an effect similar to Qt's emit?

I'm not familiar to how Qt works but judging from a few searches emit is more comparable to a ValueChangedNotifier like the implementation of the tip we currently use (or the listenable the db creates).

> How about https://pub.dev/packages/bloc ?
>
> Is it useful for us?

It might be but would need a complete rewrite of the current implementation. I already started to use throw as It can currently be used as a drop in replacement (see the history navigation where it is already used).

You can already see that I started to rewrite significant parts of the backend like the game recorder so further stuff will definetly coming...

calcitem commented:
`Calcitem`, not `Calzitem`.

I may have missed a typo when searching for my own TODO list if there is a typo here.

Leptopoda commented:
Rip. Didn't want to delete the comment sorry :)

> `Calcitem`, not `Calzitem`.
> 
> I may miss this typo when searching for my own TODO list if there is a typo here.

I already corrected the typo it's only still in an old commit and I didn't bother changing and rebasing this yet. I'll change it at the end :upside_down_face:

calcitem commented:
I see!

calcitem commented:
Hello! I've added new issues to the post below and strikethrough issues that have been dealt with or don't need to be dealt with.

https://github.com/calcitem/Sanmill/pull/455#issuecomment-1001180505

Leptopoda commented:
Hi @calcitem 
Just wanted to let you know that I don't have much time and I don't think I'll be able to continue my work any time soon. Looks more like april is when I'll have more time.

I don't have any problems validating PRs and giving feedback on code so if you want feel free to fix the remaining issues and I'll have a look. If not this probably needs to wait a bit.

calcitem commented:
OK!
